code/cheats_table/cheats_table.cpp

```diff
@@ -0,0 +1,110 @@
+	customCheats.emplace(SCP_string(bravos->cheatCode), std::move(bravos));
+
+	if (cf_exists_full("cheats.tbl", CF_TYPE_TABLES)) {
+		parse_cheat_table("cheats.tbl");
+	}
+	parse_modular_table("*-cht.tbm", parse_cheat_table);
+}
+
+void parse_cheat_table(const char* filename) {
+	try {
+		read_file_text(filename, CF_TYPE_TABLES);
+		reset_parse();
+		optional_string("#CUSTOM CHEATS");
+
+		while (optional_string("$Cheat:")) {
+			SCP_string code, msg, shipClass, shipName;
+			bool requireCheats = false;
+			bool shipSpawn = false;
+			required_string("+Code:");
+			stuff_string(code, F_RAW);
+
+			//We are limited to the buffer size used for the traditional cheats. We can still use short cheats, though.
+			if (code.length() > CHEAT_BUFFER_LEN) {
+				Warning(LOCATION, "Cheat code %s is too long. It will be cut off to maximum length (%i characters).", code.c_str(), CHEAT_BUFFER_LEN);
+				code = code.substr(0, CHEAT_BUFFER_LEN);
+			}
+
+			required_string("+Message:");
+			stuff_string(msg, F_RAW);
+
+			if (optional_string("+RequireCheats:")) {
+				stuff_boolean(&requireCheats);
+			}
+
+			if (optional_string("+SpawnShip:")) {
+				shipSpawn = true;
+				required_string("+Name:");
+				stuff_string(shipName, F_RAW);
+				required_string("+Class:");
+				stuff_string(shipClass, F_RAW);
+
+				int shipClassInfo = ship_info_lookup(shipClass.c_str());
+				if (shipClassInfo < 0) {
+					shipSpawn = false;
+					Warning(LOCATION, "There is no ship class named '%s'. The '%s' cheat won't summon a ship.", shipClass.c_str(), code.c_str());
+				}
+			}
+			
+			if (shipSpawn) {
+				std::unique_ptr<CustomCheat> shipCheat(new SpawnShipCheat(code, msg, requireCheats, shipClass, shipName));
+
+				if(customCheats.count(code) == 1) {
+					Warning(LOCATION, "A cheat for code '%s' already exists. It will be replaced.", code.c_str());
+					customCheats[code] = std::move(shipCheat);
+				} else {
+					customCheats.emplace(code, std::move(shipCheat));
+				}
+			} else {
+				std::unique_ptr<CustomCheat> cheat(new CustomCheat(code, msg, requireCheats));
+
+				if(customCheats.count(code) == 1) {
+					Warning(LOCATION, "A cheat for code '%s' already exists. It will be replaced.", code.c_str());
+					customCheats[code] = std::move(cheat);
+				} else {
+					customCheats.emplace(code, std::move(cheat));
+				}
+			}
+		}
+		required_string("#END");
+	}
+	catch (const parse::ParseException& e) {
+		mprintf(("TABLES: Unable to parse '%s'!  Error message = %s.\n", filename, e.what()));
+		return;
+	}
+}
+
+bool checkForCustomCheats(char buffer[], int buffer_length) {
+	char* check_buffer = new char[buffer_length];
+	memset (check_buffer, 0, buffer_length * sizeof(char));
+
+	// We construct a checking buffer from the original.
+	int idx = 0;
+	for (int i = 0; i < buffer_length; i++) {
+		if (buffer[i] != '\0') {
+			check_buffer[idx] = buffer[i];
+			idx++;
+		}
+	}
```

Comment:
So, why do you copy the string here?
If you strstr on buffer directly, it should work.
Also, for type safety's sake, you should pass buffer as a const char* or const char[], since you really don't want to modify it in this method

Reply:
This worked. I don't know exactly why I did it that way, but I do remember that I had a hard time and got really confused building this function 😅

Reply:
My guess is that it has to do with the null chars in the beginning of the cheat code string.
Try this, plus make ``buffer`` into a const char*.
This'll scroll check_buffer to the start of the buffer without needlessly copying it.

```suggestion
	const char* check_buffer = buffer;
	for (int i = 0; i < buffer_length; i++) {
		if (*check_buffer == '\0') {
			check_buffer++;
		}
		else {
			break;
		}
	}
```

Reply:
Ok, no, it didn't work. I remember now. Just passing the buffer doesn't work because it'll stop at the null terminator if the buffer is not full, so a cheat that's smaller than 17 characters (the max cheat length) won't work until 17 keypresses have been registered. 

I'm thinking we can just initialize the Cheat Buffer to be a string of 17 spaces or some other character.

Reply:
Either that, or you do this loop which will scroll forward to the first non-null char in the buffer

Reply:
Now I got it right. Thanks!